### PR TITLE
Add informal diagram consistency pipeline steps

### PR DESCRIPTION
**Pull request by jeanpmathes**
This PR adds a module containing steps to find trace links between diagrams and models (architecture, code). Addionally, the code model is adapted to contain more information, including references to used types and parent classes for nested classes.

### CODE REVIEW

File: stages/diagram-consistency/pom.xml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    </parent>
+
+    <groupId>edu.kit.kastel.kit.mcse.ardoco.diagramconsistency</groupId>
+    <artifactId>diagram-consistency</artifactId>
+
+    <properties>
+        <maven.compiler.source>20</maven.compiler.source>
+        <maven.compiler.target>20</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:19]: Managed by parent :)

File: stages/diagram-consistency/pom.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-kotlin</artifactId>
+            <version>2.15.2</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.ardoco.core</groupId>
+            <artifactId>common</artifactId>
+            <version>0.26.0.local-SNAPSHOT</version>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:51]:
```suggestion
            <version>${revision}</version>
```

File: stages/diagram-consistency/pom.xml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            <groupId>io.github.ardoco.core</groupId>
+            <artifactId>diagram-recognition</artifactId>
+            <version>0.26.0.local-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.ardoco.core</groupId>
+            <artifactId>tests-base</artifactId>
+            <version>0.26.0.local-SNAPSHOT</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:76]: See above :)

File: stages/diagram-consistency/pom.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
+            <version>2.15.2</version>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:41]: Version managed by parent

File: stages/diagram-consistency/pom.xml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-kotlin</artifactId>
+            <version>2.15.2</version>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:46]: Version managed by parent

File: stages/diagram-consistency/pom.xml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        <dependency>
+            <groupId>org.jgrapht</groupId>
+            <artifactId>jgrapht-ext</artifactId>
+            <version>1.5.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>RELEASE</version>
+            <scope>test</scope>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:36]: See pom.xml in other tests. Version also managed

File: stages/diagram-consistency/pom.xml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.github.ardoco.core</groupId>
+        <artifactId>stages</artifactId>
+        <version>${revision}</version>
+    </parent>
+
+    <groupId>edu.kit.kastel.kit.mcse.ardoco.diagramconsistency</groupId>
```
Comment (dfuchss) [inline stages/diagram-consistency/pom.xml:12]: Adapt Group id similar to other modules :)

**Review by dfuchss — CHANGES_REQUESTED** [review]
Some comments before CI build :)

File: framework/common/pom.xml [hunk 8]
```diff
@@ -60,5 +60,11 @@
       <artifactId>sqlite-jdbc</artifactId>
       <version>3.44.1.0</version>
     </dependency>
+      <dependency>
+          <groupId>org.jgrapht</groupId>
+          <artifactId>jgrapht-core</artifactId>
+          <version>1.5.2</version>
+          <scope>compile</scope>
+      </dependency>
```
Comment (dfuchss) [inline framework/common/pom.xml:68]: Add version of jgraph to property in outer pom.xml and use it for all modules

File: framework/common/pom.xml [hunk 9]
```diff
@@ -60,5 +60,11 @@
       <artifactId>sqlite-jdbc</artifactId>
       <version>3.44.1.0</version>
     </dependency>
+    <dependency>
+      <groupId>org.jgrapht</groupId>
+      <artifactId>jgrapht-core</artifactId>
+      <version>${jgrapht.version}</version>
+      <scope>compile</scope>
```
Comment (dfuchss) [inline framework/common/pom.xml:67]:
```suggestion
```

**Comment by dfuchss** [comment]
Two points:
* Could you run `mvn spotless:apply` once ?
* Could you add a PR to ArDoCo/Benchmark instead of changing the benchmark in this repository directly :)

**Comment by jeanpmathes** [comment]
PR in benchmark is here: https://github.com/ArDoCo/Benchmark/pull/12

**Comment by dfuchss** [comment]
Thanks :) I've updated the benchmark. Let's wait for the CI

**Comment by dfuchss** [comment]
@jeanpmathes can you look at 

```
Error:  Failures: 
Error:    DiagramLoadingTest.loadStageOne:86->EvaluationBase.getCodeModel:106 expected: not <null>
Error:    DiagramConsistencyTest.evaluateStage1VersionFinalLevenshtein:244->assertStage1Result:250 expected: <0.36> but was: <0.08>
Error:  Errors: 
Error:    DiagramConsistencyTest.evaluateAllStages:573->runAndEvaluateStage3:615 » NullPointer
[INFO] 
Error:  Tests run: 125, Failures: 2, Errors: 1, Skipped: 37
```

**Comment by jeanpmathes** [comment]
The commits might fix the first failure, but I am not sure as I can't reproduce it. For the other two problems I am still looking for a reason.

**Comment by dfuchss** [comment]
Maybe you have to clear some AppData/ArDoCo folders.

**Comment by dfuchss** [comment]
@jeanpmathes in https://github.com/ArDoCo/Core/pull/315 we beautified the JSON of the Code Model. Could you regenerate the Code Model (adapted version) and create another PR in benchmark again :) Then the review should be easy.

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramconsistency/common/Edge.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+     * @return The copy.
+     */
+    public Edge copy() {
+        return new Edge(this.label);
+    }
+
+    /**
+     * Edge labels used in the matched graphs representing diagrams and models.
+     */
+    public enum Label {
```
Comment (dfuchss) [inline framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramconsistency/common/Edge.java:41]: Extract Enum to own file :)

**Comment by jeanpmathes** [comment]
Something seems wrong with coverage calculation, it fails even before verify is done. Checking coverage locally leads to a higher value, which also makes sense because multiple of the tests run the entire pipeline that is added.

**Comment by dfuchss** [comment]
It works now :) I'll have a look at the code again but I think it will be good 👍

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramconsistency/common/TextSimilarity.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+import org.jgrapht.alg.util.UnorderedPair;
+
+import edu.kit.kastel.mcse.ardoco.core.architecture.Deterministic;
+
+/**
+ * Utility to get the similarity of two texts.
+ */
+@Deterministic
+public final class TextSimilarity {
+    private static final Map<UnorderedPair<String, String>, Double> CACHE_FOR_BASIC_FUNCTION = new java.util.LinkedHashMap<>();
```
Comment (dfuchss) [inline framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramconsistency/common/TextSimilarity.java:18]: Could we use sorted maps here. Then we could remove `@Deterministic`

File: stages/diagram-consistency/src/test/java/edu/kit/kastel/mcse/ardoco/core/diagramconsistency/evaluation/EvaluationTestBase.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    protected static CodeModel getCodeModel(DiagramProject project) {
+        File model = new File(Objects.requireNonNull(project.getSourceProject().getCodeModelDirectory())).getAbsoluteFile();
+        CodeExtractor extractor = new AllLanguagesExtractor(new CodeItemRepository(), model.getAbsolutePath());
+
+        CodeModel codeModel = extractor.readInCodeModel();
+
+        if (codeModel == null) {
+            String repository = project.getSourceProject().getCodeRepository();
```
Comment (dfuchss) [inline stages/diagram-consistency/src/test/java/edu/kit/kastel/mcse/ardoco/core/diagramconsistency/evaluation/EvaluationTestBase.java:110]: I'll have a look here. The loading should be handled by the enum (internally)

**Review by dfuchss — APPROVED** [review]
In summary, it looks good to me. Thanks a lot.

File: stages/model-provider/src/test/resources/interface/edu/AClass.java [hunk 13]
```diff
@@ -4,6 +4,22 @@
 // import edu.drei.OtherInterface;
```
Comment (Gram21) [inline stages/model-provider/src/test/resources/interface/edu/AClass.java:4]: Even if it was there before, maybe remove this commented out import

**Review by Gram21 — APPROVED** [review]

**Comment by dfuchss** [comment]
@jeanpmathes thanks for the PR. Well done :) I'll merge it now !

**Pull request merged**